Library/Homebrew/unpack_strategy.rb

```diff
@@ -0,0 +1,340 @@
+class UnpackStrategy
+  # length of the longest regex (currently TarUnpackStrategy)
+  MAX_MAGIC_NUMBER_LENGTH = 262
+  private_constant :MAX_MAGIC_NUMBER_LENGTH
+
+  def self.strategies
+    @strategies ||= [
+      JarUnpackStrategy,
+      ZipUnpackStrategy,
+      XarUnpackStrategy,
+      XzUnpackStrategy,
+      TarUnpackStrategy,
+      Bzip2UnpackStrategy,
+      GzipUnpackStrategy,
+      LzipUnpackStrategy,
+      GitUnpackStrategy,
+      MercurialUnpackStrategy,
+      SubversionUnpackStrategy,
+      CvsUnpackStrategy,
+      FossilUnpackStrategy,
+      BazaarUnpackStrategy,
+      P7ZipUnpackStrategy,
+      RarUnpackStrategy,
+      LhaUnpackStrategy,
+    ].freeze
+  end
+  private_class_method :strategies
+
+  def self.detect(path)
+    magic_number = if path.directory?
+      ""
+    else
+      File.binread(path, MAX_MAGIC_NUMBER_LENGTH)
+    end
+
+    strategy = strategies.detect do |s|
+      s.can_extract?(path: path, magic_number: magic_number)
+    end
+
+    # This is so that bad files produce good error messages.
+    strategy ||= case path.extname
+    when ".tar.gz", ".tgz", ".tar.bz2", ".tbz", ".tar.xz"
+      TarUnpackStrategy
+    when ".zip"
+      ZipUnpackStrategy
+    else
+      UncompressedUnpackStrategy
+    end
+
+    strategy.new(path)
+  end
+
+  attr_reader :path
+
+  def initialize(path)
+    @path = Pathname(path).expand_path
+  end
+
+  def extract(to: nil, basename: nil)
+    unpack_dir = Pathname(to || Dir.pwd).expand_path
+    unpack_dir.mkpath
+    extract_to_dir(unpack_dir, basename: basename)
+  end
+
+  def pipe_to_tar(tool, unpack_dir)
+    Utils.popen_read(tool, "-dc", path) do |rd|
+      Utils.popen_write("tar", "-x", "-f", "-", "-C", unpack_dir) do |wr|
+        buf = ""
+        wr.write(buf) while rd.read(16384, buf)
```

Comment:
Can `16384` be made a constant?

Reply:
This should be possible without a buffer by redirecting the first output to the second input.